File: library/builtins/primitives/list.rs
use crate::*;
use wipple::*;

#[derive(TypeInfo, Debug, Clone)]
pub struct List {
    pub items: Vec<Value>,
    pub location: Option<SourceLocation>,
}

impl Primitive for List {}

impl List {
    pub fn new(items: Vec<Value>) -> Self {
        List::new_located(items, None)
    }

    pub fn new_located(items: Vec<Value>, location: Option<SourceLocation>) -> Self {
        List {
            items: items.to_vec(),
            location,
        }
    }
}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    env.set_variable("List", Value::of(Trait::of::<List>()));

    // List == Evaluate
    env.add_relation_between(stack, |list: List| {
        EvaluateFn::new(move |env, stack| {
            let mut stack = stack.clone();
            stack
                .evaluation_mut()
                .queue_location(list.location.as_ref());

            let operators = list.find_operators(env, &stack)?;

            if let Some(parsed) = list.parse_operators(operators, env, &stack)? {
                return Ok(parsed);
            }

            // Reduce the list as a series of function calls

            let mut result = match list.items.first() {
                Some(value) => value.evaluate(env, &stack)?.into_owned(),
                None => {
                    // Empty list evaluates to itself
                    return Ok(Value::of(list.clone()));
                }
            };

            for item in list.items.clone().into_iter().skip(1) {
                result = result.call_with(item, env, &stack)?;
            }

            Ok(result)
        })
    })?;

    // List == Text
    env.add_relation_between_with(stack, |list: List, env, stack| {
        let mut items = String::new();

        let count = list.items.len();

        for (index, item) in list.items.into_iter().enumerate() {
            let text = item.format(env, stack)?;
            items.push_str(&text);

            if index != count - 1 {
                items.push(' ');
            }
        }

        Ok(Text::new(format!("({})", items)))
    })?;

    Ok(())
}

File: Library/Sources/Traits/List.swift
import Wipple

public struct List: RawRepresentable, Primitive {
    public var rawValue: [Value]

    public init(rawValue: [Value]) {
        self.rawValue = rawValue
    }
}

public extension Value {
    init<Items: Collection>(_ list: Items) where Items.Element == Value {
        if list.count == 1 {
            self = list.first!
        } else {
            self = Value(List(rawValue: Array(list)))
        }
    }
}

internal func setupList(_ env: Env, _ stack: Stack) throws {
    try env.setVariable("List", to: Value(Trait(List.self)), stack)

    // List == Evaluate
    try env.addRelation(from: List.self, to: Evaluate.self, stack) { list, value in
        Evaluate { env, stack in
            var stack = stack
            stack.diagnostics.queue(value.location)

            // Parse operators contained within the list
            if let parsedValue = try list.parseOperators(env, stack) {
                return parsedValue
            }

            // Reduce the list as a series of function calls

            guard var result = try list.rawValue.first?.evaluate(env, stack) else {
                // Empty list evaluates to itself
                return Value(list)
            }

            for item in list.rawValue.dropFirst() {
                result = try result.call(with: item, env, stack)
            }

            return result
        }
    }

    // List == Interpolate
    try env.addRelation(from: List.self, to: Interpolate.self, stack) { list in
        Interpolate { direct, env, stack in
            let items = try list.rawValue.map { item in
                try item.interpolate(direct: direct, env, stack)
            }

            return Value(List(rawValue: items))
        }
    }

    // List == Text
    try env.addRelation(from: List.self, to: Text.self, stack) { list, _, stack in
        let items = try list.rawValue
            .map { try $0.format(env, stack) }
            .joined(separator: " ")

        return Text(rawValue: "(\(items))")
    }
}
